// @syncfusion/ej2-ng-charts/src/sparkline/rangebandsettings.directive 2.js
var __extends = (this && this.__extends) || (function () {
    var extendStatics = Object.setPrototypeOf ||
        ({ __proto__: [] } instanceof Array && function (d, b) { d.__proto__ = b; }) ||
        function (d, b) { for (var p in b) if (b.hasOwnProperty(p)) d[p] = b[p]; };
    return function (d, b) {
        extendStatics(d, b);
        function __() { this.constructor = d; }
        d.prototype = b === null ? Object.create(b) : (__.prototype = b.prototype, new __());
    };
})();
import { Directive, ViewContainerRef, ContentChildren } from '@angular/core';
import { ComplexBase, ArrayBase, setValue } from '@syncfusion/ej2-ng-base';
var input = ['color', 'endRange', 'opacity', 'startRange'];
var outputs = ['color', 'endRange', 'opacity', 'startRange'];
var RangeBandSettingDirective = /** @class */ (function (_super) {
    __extends(RangeBandSettingDirective, _super);
    function RangeBandSettingDirective(viewContainerRef) {
        var _this = _super.call(this) || this;
        _this.viewContainerRef = viewContainerRef;
        setValue('currentInstance', _this, _this.viewContainerRef);
        return _this;
        //this.registerEvents(outputs);
    }
    RangeBandSettingDirective.decorators = [
        { type: Directive, args: [{
                    selector: 'e-rangeBandSettings>e-rangeBandSetting',
                    inputs: input,
                    queries: {}
                },] },
    ];
    /** @nocollapse */
    RangeBandSettingDirective.ctorParameters = function () { return [
        { type: ViewContainerRef, },
    ]; };
    return RangeBandSettingDirective;
}(ComplexBase));
export { RangeBandSettingDirective };
/**
 * RangeBandSetting Array Directive
 * @private
 */
var RangeBandSettingsDirective = /** @class */ (function (_super) {
    __extends(RangeBandSettingsDirective, _super);
    function RangeBandSettingsDirective() {
        return _super.call(this, 'rangebandsettings') || this;
    }
    RangeBandSettingsDirective.decorators = [
        { type: Directive, args: [{
                    selector: 'ejs-sparkline>e-rangeBandSettings',
                    queries: {
                        children: new ContentChildren(RangeBandSettingDirective)
                    },
                },] },
    ];
    /** @nocollapse */
    RangeBandSettingsDirective.ctorParameters = function () { return []; };
    return RangeBandSettingsDirective;
}(ArrayBase));
export { RangeBandSettingsDirective };
