// @syncfusion/ej2-pdf-export/src/implementation/graphics/enum 2.js
define(["require", "exports"], function (require, exports) {
    "use strict";
    Object.defineProperty(exports, "__esModule", { value: true });
    var PdfHorizontalAlignment;
    (function (PdfHorizontalAlignment) {
        PdfHorizontalAlignment[PdfHorizontalAlignment["Left"] = 0] = "Left";
        PdfHorizontalAlignment[PdfHorizontalAlignment["Center"] = 1] = "Center";
        PdfHorizontalAlignment[PdfHorizontalAlignment["Right"] = 2] = "Right";
    })(PdfHorizontalAlignment = exports.PdfHorizontalAlignment || (exports.PdfHorizontalAlignment = {}));
    var PdfVerticalAlignment;
    (function (PdfVerticalAlignment) {
        PdfVerticalAlignment[PdfVerticalAlignment["Top"] = 0] = "Top";
        PdfVerticalAlignment[PdfVerticalAlignment["Middle"] = 1] = "Middle";
        PdfVerticalAlignment[PdfVerticalAlignment["Bottom"] = 2] = "Bottom";
    })(PdfVerticalAlignment = exports.PdfVerticalAlignment || (exports.PdfVerticalAlignment = {}));
    var PdfTextAlignment;
    (function (PdfTextAlignment) {
        PdfTextAlignment[PdfTextAlignment["Left"] = 0] = "Left";
        PdfTextAlignment[PdfTextAlignment["Center"] = 1] = "Center";
        PdfTextAlignment[PdfTextAlignment["Right"] = 2] = "Right";
        PdfTextAlignment[PdfTextAlignment["Justify"] = 3] = "Justify";
    })(PdfTextAlignment = exports.PdfTextAlignment || (exports.PdfTextAlignment = {}));
    var TextRenderingMode;
    (function (TextRenderingMode) {
        TextRenderingMode[TextRenderingMode["Fill"] = 0] = "Fill";
        TextRenderingMode[TextRenderingMode["Stroke"] = 1] = "Stroke";
        TextRenderingMode[TextRenderingMode["FillStroke"] = 2] = "FillStroke";
        TextRenderingMode[TextRenderingMode["None"] = 3] = "None";
        TextRenderingMode[TextRenderingMode["ClipFlag"] = 4] = "ClipFlag";
        TextRenderingMode[TextRenderingMode["ClipFill"] = 4] = "ClipFill";
        TextRenderingMode[TextRenderingMode["ClipStroke"] = 5] = "ClipStroke";
        TextRenderingMode[TextRenderingMode["ClipFillStroke"] = 6] = "ClipFillStroke";
        TextRenderingMode[TextRenderingMode["Clip"] = 7] = "Clip";
    })(TextRenderingMode = exports.TextRenderingMode || (exports.TextRenderingMode = {}));
    var PdfLineJoin;
    (function (PdfLineJoin) {
        PdfLineJoin[PdfLineJoin["Miter"] = 0] = "Miter";
        PdfLineJoin[PdfLineJoin["Round"] = 1] = "Round";
        PdfLineJoin[PdfLineJoin["Bevel"] = 2] = "Bevel";
    })(PdfLineJoin = exports.PdfLineJoin || (exports.PdfLineJoin = {}));
    var PdfLineCap;
    (function (PdfLineCap) {
        PdfLineCap[PdfLineCap["Flat"] = 0] = "Flat";
        PdfLineCap[PdfLineCap["Round"] = 1] = "Round";
        PdfLineCap[PdfLineCap["Square"] = 2] = "Square";
    })(PdfLineCap = exports.PdfLineCap || (exports.PdfLineCap = {}));
    var PdfDashStyle;
    (function (PdfDashStyle) {
        PdfDashStyle[PdfDashStyle["Solid"] = 0] = "Solid";
        PdfDashStyle[PdfDashStyle["Dash"] = 1] = "Dash";
        PdfDashStyle[PdfDashStyle["Dot"] = 2] = "Dot";
        PdfDashStyle[PdfDashStyle["DashDot"] = 3] = "DashDot";
        PdfDashStyle[PdfDashStyle["DashDotDot"] = 4] = "DashDotDot";
        PdfDashStyle[PdfDashStyle["Custom"] = 5] = "Custom";
    })(PdfDashStyle = exports.PdfDashStyle || (exports.PdfDashStyle = {}));
    var PdfFillMode;
    (function (PdfFillMode) {
        PdfFillMode[PdfFillMode["Winding"] = 0] = "Winding";
        PdfFillMode[PdfFillMode["Alternate"] = 1] = "Alternate";
    })(PdfFillMode = exports.PdfFillMode || (exports.PdfFillMode = {}));
    var PdfColorSpace;
    (function (PdfColorSpace) {
        PdfColorSpace[PdfColorSpace["Rgb"] = 0] = "Rgb";
        PdfColorSpace[PdfColorSpace["Cmyk"] = 1] = "Cmyk";
        PdfColorSpace[PdfColorSpace["GrayScale"] = 2] = "GrayScale";
        PdfColorSpace[PdfColorSpace["Indexed"] = 3] = "Indexed";
    })(PdfColorSpace = exports.PdfColorSpace || (exports.PdfColorSpace = {}));
    var PdfBlendMode;
    (function (PdfBlendMode) {
        PdfBlendMode[PdfBlendMode["Normal"] = 0] = "Normal";
        PdfBlendMode[PdfBlendMode["Multiply"] = 1] = "Multiply";
        PdfBlendMode[PdfBlendMode["Screen"] = 2] = "Screen";
        PdfBlendMode[PdfBlendMode["Overlay"] = 3] = "Overlay";
        PdfBlendMode[PdfBlendMode["Darken"] = 4] = "Darken";
        PdfBlendMode[PdfBlendMode["Lighten"] = 5] = "Lighten";
        PdfBlendMode[PdfBlendMode["ColorDodge"] = 6] = "ColorDodge";
        PdfBlendMode[PdfBlendMode["ColorBurn"] = 7] = "ColorBurn";
        PdfBlendMode[PdfBlendMode["HardLight"] = 8] = "HardLight";
        PdfBlendMode[PdfBlendMode["SoftLight"] = 9] = "SoftLight";
        PdfBlendMode[PdfBlendMode["Difference"] = 10] = "Difference";
        PdfBlendMode[PdfBlendMode["Exclusion"] = 11] = "Exclusion";
        PdfBlendMode[PdfBlendMode["Hue"] = 12] = "Hue";
        PdfBlendMode[PdfBlendMode["Saturation"] = 13] = "Saturation";
        PdfBlendMode[PdfBlendMode["Color"] = 14] = "Color";
        PdfBlendMode[PdfBlendMode["Luminosity"] = 15] = "Luminosity";
    })(PdfBlendMode = exports.PdfBlendMode || (exports.PdfBlendMode = {}));
    var PdfGraphicsUnit;
    (function (PdfGraphicsUnit) {
        PdfGraphicsUnit[PdfGraphicsUnit["Centimeter"] = 0] = "Centimeter";
        PdfGraphicsUnit[PdfGraphicsUnit["Pica"] = 1] = "Pica";
        PdfGraphicsUnit[PdfGraphicsUnit["Pixel"] = 2] = "Pixel";
        PdfGraphicsUnit[PdfGraphicsUnit["Point"] = 3] = "Point";
        PdfGraphicsUnit[PdfGraphicsUnit["Inch"] = 4] = "Inch";
        PdfGraphicsUnit[PdfGraphicsUnit["Document"] = 5] = "Document";
        PdfGraphicsUnit[PdfGraphicsUnit["Millimeter"] = 6] = "Millimeter";
    })(PdfGraphicsUnit = exports.PdfGraphicsUnit || (exports.PdfGraphicsUnit = {}));
    var PdfGridImagePosition;
    (function (PdfGridImagePosition) {
        PdfGridImagePosition[PdfGridImagePosition["Fit"] = 0] = "Fit";
        PdfGridImagePosition[PdfGridImagePosition["Center"] = 1] = "Center";
        PdfGridImagePosition[PdfGridImagePosition["Stretch"] = 2] = "Stretch";
        PdfGridImagePosition[PdfGridImagePosition["Tile"] = 3] = "Tile";
    })(PdfGridImagePosition = exports.PdfGridImagePosition || (exports.PdfGridImagePosition = {}));
});
