// @syncfusion/ej2-pdf-export/src/implementation/graphics/pdf-margins.js
define(["require", "exports"], function (require, exports) {
    "use strict";
    Object.defineProperty(exports, "__esModule", { value: true });
    var PdfMargins = (function () {
        function PdfMargins() {
            this.pdfMargin = 40.0;
            this.setMargins(this.pdfMargin);
        }
        Object.defineProperty(PdfMargins.prototype, "left", {
            get: function () {
                return this.leftMargin;
            },
            set: function (value) {
                this.leftMargin = value;
            },
            enumerable: true,
            configurable: true
        });
        Object.defineProperty(PdfMargins.prototype, "top", {
            get: function () {
                return this.topMargin;
            },
            set: function (value) {
                this.topMargin = value;
            },
            enumerable: true,
            configurable: true
        });
        Object.defineProperty(PdfMargins.prototype, "right", {
            get: function () {
                return this.rightMargin;
            },
            set: function (value) {
                this.rightMargin = value;
            },
            enumerable: true,
            configurable: true
        });
        Object.defineProperty(PdfMargins.prototype, "bottom", {
            get: function () {
                return this.bottomMargin;
            },
            set: function (value) {
                this.bottomMargin = value;
            },
            enumerable: true,
            configurable: true
        });
        Object.defineProperty(PdfMargins.prototype, "all", {
            set: function (value) {
                this.setMargins(value);
            },
            enumerable: true,
            configurable: true
        });
        PdfMargins.prototype.setMargins = function (margin1, margin2, margin3, margin4) {
            if (typeof margin2 === 'undefined') {
                this.leftMargin = this.topMargin = this.rightMargin = this.bottomMargin = margin1;
            }
            else {
                if (typeof margin3 === 'undefined') {
                    this.leftMargin = this.rightMargin = margin1;
                    this.bottomMargin = this.topMargin = margin2;
                }
                else {
                    this.leftMargin = margin1;
                    this.topMargin = margin2;
                    this.rightMargin = margin3;
                    this.bottomMargin = margin4;
                }
            }
        };
        PdfMargins.prototype.clone = function () {
            return this;
        };
        return PdfMargins;
    }());
    exports.PdfMargins = PdfMargins;
});
