// @syncfusion/ej2-pdf-export/src/implementation/graphics/unit-convertor 2.js
define(["require", "exports"], function (require, exports) {
    "use strict";
    Object.defineProperty(exports, "__esModule", { value: true });
    var PdfUnitConverter = (function () {
        function PdfUnitConverter(dpi) {
            this.updateProportionsHelper(dpi);
        }
        PdfUnitConverter.prototype.convertUnits = function (value, from, to) {
            return this.convertFromPixels(this.convertToPixels(value, from), to);
        };
        PdfUnitConverter.prototype.convertToPixels = function (value, from) {
            var index = from;
            var result = (value * this.proportions[index]);
            return result;
        };
        PdfUnitConverter.prototype.convertFromPixels = function (value, to) {
            var index = to;
            var result = (value / this.proportions[index]);
            return result;
        };
        PdfUnitConverter.prototype.updateProportionsHelper = function (pixelPerInch) {
            this.proportions = [
                pixelPerInch / 2.54,
                pixelPerInch / 6.0,
                1,
                pixelPerInch / 72.0,
                pixelPerInch,
                pixelPerInch / 300.0,
                pixelPerInch / 25.4
            ];
        };
        PdfUnitConverter.horizontalResolution = 96;
        PdfUnitConverter.verticalResolution = 96;
        return PdfUnitConverter;
    }());
    exports.PdfUnitConverter = PdfUnitConverter;
});
