// @syncfusion/ej2-pdf-export/src/implementation/pages/pdf-section-page-collection.js
define(["require", "exports"], function (require, exports) {
    "use strict";
    Object.defineProperty(exports, "__esModule", { value: true });
    var PdfSectionPageCollection = (function () {
        function PdfSectionPageCollection(section) {
            this.pdfSection = null;
            if (section == null) {
                throw Error('ArgumentNullException("section")');
            }
            this.section = section;
        }
        Object.defineProperty(PdfSectionPageCollection.prototype, "section", {
            get: function () {
                return this.pdfSection;
            },
            set: function (value) {
                this.pdfSection = value;
            },
            enumerable: true,
            configurable: true
        });
        PdfSectionPageCollection.prototype.contains = function (page) {
            return this.section.contains(page);
        };
        PdfSectionPageCollection.prototype.remove = function (page) {
            this.section.remove(page);
        };
        PdfSectionPageCollection.prototype.add = function () {
            return this.section.add();
        };
        return PdfSectionPageCollection;
    }());
    exports.PdfSectionPageCollection = PdfSectionPageCollection;
});
