// @syncfusion/ej2-pdf-export/src/implementation/pages/pdf-section-templates 2.js
var __extends = (this && this.__extends) || (function () {
    var extendStatics = Object.setPrototypeOf ||
        ({ __proto__: [] } instanceof Array && function (d, b) { d.__proto__ = b; }) ||
        function (d, b) { for (var p in b) if (b.hasOwnProperty(p)) d[p] = b[p]; };
    return function (d, b) {
        extendStatics(d, b);
        function __() { this.constructor = d; }
        d.prototype = b === null ? Object.create(b) : (__.prototype = b.prototype, new __());
    };
})();
define(["require", "exports", "./../document/pdf-document-template"], function (require, exports, pdf_document_template_1) {
    "use strict";
    Object.defineProperty(exports, "__esModule", { value: true });
    var PdfSectionTemplate = (function (_super) {
        __extends(PdfSectionTemplate, _super);
        function PdfSectionTemplate() {
            var _this = _super.call(this) || this;
            _this.leftValue = _this.topValue = _this.rightValue = _this.bottomValue = _this.stampValue = true;
            return _this;
        }
        Object.defineProperty(PdfSectionTemplate.prototype, "applyDocumentLeftTemplate", {
            get: function () {
                return this.leftValue;
            },
            set: function (value) {
                this.leftValue = value;
            },
            enumerable: true,
            configurable: true
        });
        Object.defineProperty(PdfSectionTemplate.prototype, "applyDocumentTopTemplate", {
            get: function () {
                return this.topValue;
            },
            set: function (value) {
                this.topValue = value;
            },
            enumerable: true,
            configurable: true
        });
        Object.defineProperty(PdfSectionTemplate.prototype, "applyDocumentRightTemplate", {
            get: function () {
                return this.rightValue;
            },
            set: function (value) {
                this.rightValue = value;
            },
            enumerable: true,
            configurable: true
        });
        Object.defineProperty(PdfSectionTemplate.prototype, "applyDocumentBottomTemplate", {
            get: function () {
                return this.bottomValue;
            },
            set: function (value) {
                this.bottomValue = value;
            },
            enumerable: true,
            configurable: true
        });
        Object.defineProperty(PdfSectionTemplate.prototype, "applyDocumentStamps", {
            get: function () {
                return this.stampValue;
            },
            set: function (value) {
                this.stampValue = value;
            },
            enumerable: true,
            configurable: true
        });
        return PdfSectionTemplate;
    }(pdf_document_template_1.PdfDocumentTemplate));
    exports.PdfSectionTemplate = PdfSectionTemplate;
});
