// @syncfusion/ej2-ng-charts/src/smithchart/smithchart.component 2.js
var __extends = (this && this.__extends) || (function () {
    var extendStatics = Object.setPrototypeOf ||
        ({ __proto__: [] } instanceof Array && function (d, b) { d.__proto__ = b; }) ||
        function (d, b) { for (var p in b) if (b.hasOwnProperty(p)) d[p] = b[p]; };
    return function (d, b) {
        extendStatics(d, b);
        function __() { this.constructor = d; }
        d.prototype = b === null ? Object.create(b) : (__.prototype = b.prototype, new __());
    };
})();
import { Component, ElementRef, ViewContainerRef, ChangeDetectionStrategy, Renderer2, ContentChild, Optional } from '@angular/core';
import { ComponentBase, applyMixins, setValue } from '@syncfusion/ej2-ng-base';
import { Smithchart, SmithchartLegend, TooltipRender } from '@syncfusion/ej2-charts';
import { SmithchartSeriesCollectionDirective } from './series.directive';
export var inputs = ['background', 'border', 'elementSpacing', 'enablePersistence', 'enableRtl', 'font', 'height', 'horizontalAxis', 'isResponsive', 'legendSettings', 'locale', 'margin', 'radialAxis', 'radius', 'renderType', 'series', 'theme', 'title', 'width'];
export var outputs = ['animationComplete', 'axisLabelRender', 'beforePrint', 'legendRender', 'load', 'loaded', 'seriesRender', 'subtitleRender', 'textRender', 'titleRender'];
export var twoWays = [''];
/**
 * Smithchart Component
 * ```html
 * <ejs-smithchart></ejs-smithchart>
 * ```
 */
var SmithchartComponent = /** @class */ (function (_super) {
    __extends(SmithchartComponent, _super);
    function SmithchartComponent(ngEle, srenderer, viewContainerRef, smithchartLegend, tooltipRender) {
        var _this = _super.call(this) || this;
        _this.ngEle = ngEle;
        _this.srenderer = srenderer;
        _this.viewContainerRef = viewContainerRef;
        _this.tags = ['series'];
        _this.element = _this.ngEle.nativeElement;
        _this.injectedModules = _this.injectedModules || [];
        for (var i = 3; i < arguments.length; i++) {
            var injectedObj = arguments[i];
            if (injectedObj) {
                _this.injectedModules.push(injectedObj);
            }
        }
        _this.registerEvents(outputs);
        _this.addTwoWay.call(_this, twoWays);
        setValue('currentInstance', _this, _this.viewContainerRef);
        return _this;
    }
    SmithchartComponent.prototype.ngOnInit = function () {
    };
    SmithchartComponent.prototype.ngAfterViewInit = function () {
    };
    SmithchartComponent.prototype.ngOnDestroy = function () {
    };
    SmithchartComponent.prototype.ngAfterContentChecked = function () {
    };
    SmithchartComponent.decorators = [
        { type: Component, args: [{
                    selector: 'ejs-smithchart',
                    inputs: inputs,
                    outputs: outputs,
                    template: '',
                    changeDetection: ChangeDetectionStrategy.OnPush,
                    queries: {
                        childSeries: new ContentChild(SmithchartSeriesCollectionDirective)
                    }
                },] },
    ];
    /** @nocollapse */
    SmithchartComponent.ctorParameters = function () { return [
        { type: ElementRef, },
        { type: Renderer2, },
        { type: ViewContainerRef, },
        { type: SmithchartLegend, decorators: [{ type: Optional },] },
        { type: TooltipRender, decorators: [{ type: Optional },] },
    ]; };
    return SmithchartComponent;
}(Smithchart));
export { SmithchartComponent };
applyMixins(SmithchartComponent, [ComponentBase]);
export var SmithchartLegendService = { provide: SmithchartLegend, useValue: SmithchartLegend };
export var TooltipRenderService = { provide: TooltipRender, useValue: TooltipRender };
