// @syncfusion/ej2-pdf-export/src/implementation/structured-elements/grid/styles/pdf-borders 2.js
define(["require", "exports", "./../../../graphics/pdf-pen", "./../../../graphics/enum", "./../../../graphics/pdf-color"], function (require, exports, pdf_pen_1, enum_1, pdf_color_1) {
    "use strict";
    Object.defineProperty(exports, "__esModule", { value: true });
    var PdfBorders = (function () {
        function PdfBorders() {
            var defaultBorderPenLeft = new pdf_pen_1.PdfPen(new pdf_color_1.PdfColor(0, 0, 0));
            defaultBorderPenLeft.dashStyle = enum_1.PdfDashStyle.Solid;
            var defaultBorderPenRight = new pdf_pen_1.PdfPen(new pdf_color_1.PdfColor(0, 0, 0));
            defaultBorderPenRight.dashStyle = enum_1.PdfDashStyle.Solid;
            var defaultBorderPenTop = new pdf_pen_1.PdfPen(new pdf_color_1.PdfColor(0, 0, 0));
            defaultBorderPenTop.dashStyle = enum_1.PdfDashStyle.Solid;
            var defaultBorderPenBottom = new pdf_pen_1.PdfPen(new pdf_color_1.PdfColor(0, 0, 0));
            defaultBorderPenBottom.dashStyle = enum_1.PdfDashStyle.Solid;
            this.leftPen = defaultBorderPenLeft;
            this.rightPen = defaultBorderPenRight;
            this.topPen = defaultBorderPenTop;
            this.bottomPen = defaultBorderPenBottom;
        }
        Object.defineProperty(PdfBorders.prototype, "left", {
            get: function () {
                return this.leftPen;
            },
            set: function (value) {
                this.leftPen = value;
            },
            enumerable: true,
            configurable: true
        });
        Object.defineProperty(PdfBorders.prototype, "right", {
            get: function () {
                return this.rightPen;
            },
            set: function (value) {
                this.rightPen = value;
            },
            enumerable: true,
            configurable: true
        });
        Object.defineProperty(PdfBorders.prototype, "top", {
            get: function () {
                return this.topPen;
            },
            set: function (value) {
                this.topPen = value;
            },
            enumerable: true,
            configurable: true
        });
        Object.defineProperty(PdfBorders.prototype, "bottom", {
            get: function () {
                return this.bottomPen;
            },
            set: function (value) {
                this.bottomPen = value;
            },
            enumerable: true,
            configurable: true
        });
        Object.defineProperty(PdfBorders.prototype, "all", {
            set: function (value) {
                this.leftPen = this.rightPen = this.topPen = this.bottomPen = value;
            },
            enumerable: true,
            configurable: true
        });
        Object.defineProperty(PdfBorders.prototype, "isAll", {
            get: function () {
                return ((this.leftPen === this.rightPen) && (this.leftPen === this.topPen) && (this.leftPen === this.bottomPen));
            },
            enumerable: true,
            configurable: true
        });
        Object.defineProperty(PdfBorders, "default", {
            get: function () {
                return new PdfBorders();
            },
            enumerable: true,
            configurable: true
        });
        return PdfBorders;
    }());
    exports.PdfBorders = PdfBorders;
    var PdfPaddings = (function () {
        function PdfPaddings(left, right, top, bottom) {
            if (typeof left === 'undefined') {
                this.leftPad = this.rightPad = 5.76;
                this.bottom = this.topPad = 0.5;
            }
            else {
                this.leftPad = left;
                this.rightPad = right;
                this.topPad = top;
                this.bottomPad = bottom;
            }
        }
        Object.defineProperty(PdfPaddings.prototype, "left", {
            get: function () {
                return this.leftPad;
            },
            set: function (value) {
                this.leftPad = value;
            },
            enumerable: true,
            configurable: true
        });
        Object.defineProperty(PdfPaddings.prototype, "right", {
            get: function () {
                return this.rightPad;
            },
            set: function (value) {
                this.rightPad = value;
            },
            enumerable: true,
            configurable: true
        });
        Object.defineProperty(PdfPaddings.prototype, "top", {
            get: function () {
                return this.topPad;
            },
            set: function (value) {
                this.topPad = value;
            },
            enumerable: true,
            configurable: true
        });
        Object.defineProperty(PdfPaddings.prototype, "bottom", {
            get: function () {
                return this.bottomPad;
            },
            set: function (value) {
                this.bottomPad = value;
            },
            enumerable: true,
            configurable: true
        });
        Object.defineProperty(PdfPaddings.prototype, "all", {
            set: function (value) {
                this.leftPad = this.rightPad = this.topPad = this.bottomPad = value;
            },
            enumerable: true,
            configurable: true
        });
        return PdfPaddings;
    }());
    exports.PdfPaddings = PdfPaddings;
});
