// @syncfusion/ej2-ng-charts/src/range-navigator/series.directive 2.js
var __extends = (this && this.__extends) || (function () {
    var extendStatics = Object.setPrototypeOf ||
        ({ __proto__: [] } instanceof Array && function (d, b) { d.__proto__ = b; }) ||
        function (d, b) { for (var p in b) if (b.hasOwnProperty(p)) d[p] = b[p]; };
    return function (d, b) {
        extendStatics(d, b);
        function __() { this.constructor = d; }
        d.prototype = b === null ? Object.create(b) : (__.prototype = b.prototype, new __());
    };
})();
import { Directive, ViewContainerRef, ContentChildren } from '@angular/core';
import { ComplexBase, ArrayBase, setValue } from '@syncfusion/ej2-ng-base';
var input = ['animation', 'border', 'dashArray', 'dataSource', 'fill', 'query', 'type', 'width', 'xName', 'yName'];
var outputs = ['animation', 'border', 'dashArray', 'dataSource', 'fill', 'query', 'type', 'width', 'xName', 'yName'];
/**
 * RangenavigatorSeries Directive
 * ```html
 * <e-rangenavigator-series-collection>
 * <e-rangenavigator-series></e-rangenavigator-series>
 * </e-rangenavigator-series-collection>
 * ```
 */
var RangenavigatorSeriesDirective = /** @class */ (function (_super) {
    __extends(RangenavigatorSeriesDirective, _super);
    function RangenavigatorSeriesDirective(viewContainerRef) {
        var _this = _super.call(this) || this;
        _this.viewContainerRef = viewContainerRef;
        setValue('currentInstance', _this, _this.viewContainerRef);
        return _this;
        //this.registerEvents(outputs);
    }
    RangenavigatorSeriesDirective.decorators = [
        { type: Directive, args: [{
                    selector: 'e-rangenavigator-series-collection>e-rangenavigator-series',
                    inputs: input,
                    queries: {}
                },] },
    ];
    /** @nocollapse */
    RangenavigatorSeriesDirective.ctorParameters = function () { return [
        { type: ViewContainerRef, },
    ]; };
    return RangenavigatorSeriesDirective;
}(ComplexBase));
export { RangenavigatorSeriesDirective };
/**
 * RangenavigatorSeries Array Directive
 * @private
 */
var RangenavigatorSeriesCollectionDirective = /** @class */ (function (_super) {
    __extends(RangenavigatorSeriesCollectionDirective, _super);
    function RangenavigatorSeriesCollectionDirective() {
        return _super.call(this, 'series') || this;
    }
    RangenavigatorSeriesCollectionDirective.decorators = [
        { type: Directive, args: [{
                    selector: 'ej-rangenavigator>e-rangenavigator-series-collection',
                    queries: {
                        children: new ContentChildren(RangenavigatorSeriesDirective)
                    },
                },] },
    ];
    /** @nocollapse */
    RangenavigatorSeriesCollectionDirective.ctorParameters = function () { return []; };
    return RangenavigatorSeriesCollectionDirective;
}(ArrayBase));
export { RangenavigatorSeriesCollectionDirective };
