// @syncfusion/ej2-ng-charts/src/chart/categories.directive.js
var __extends = (this && this.__extends) || (function () {
    var extendStatics = Object.setPrototypeOf ||
        ({ __proto__: [] } instanceof Array && function (d, b) { d.__proto__ = b; }) ||
        function (d, b) { for (var p in b) if (b.hasOwnProperty(p)) d[p] = b[p]; };
    return function (d, b) {
        extendStatics(d, b);
        function __() { this.constructor = d; }
        d.prototype = b === null ? Object.create(b) : (__.prototype = b.prototype, new __());
    };
})();
import { Directive, ViewContainerRef, ContentChildren } from '@angular/core';
import { ComplexBase, ArrayBase, setValue } from '@syncfusion/ej2-ng-base';
var input = ['end', 'maximumTextWidth', 'start', 'text'];
var outputs = ['end', 'maximumTextWidth', 'start', 'text'];
/**
 * MultiLevelLabels Directive
 * ```html
 * <e-multilevellabels>
 * <e-multilevellabel>
 * <e-Categories>
 * <e-Category>
 * </e-Category>
 * </e-Categories>
 * </e-multilevellabel>
 * </e-multilevellabels>
 * ```
 */
var CategoryDirective = /** @class */ (function (_super) {
    __extends(CategoryDirective, _super);
    function CategoryDirective(viewContainerRef) {
        var _this = _super.call(this) || this;
        _this.viewContainerRef = viewContainerRef;
        setValue('currentInstance', _this, _this.viewContainerRef);
        return _this;
        //this.registerEvents(outputs);
    }
    CategoryDirective.decorators = [
        { type: Directive, args: [{
                    selector: 'e-multilevellabel>e-categories>e-category',
                    inputs: input,
                    queries: {}
                },] },
    ];
    /** @nocollapse */
    CategoryDirective.ctorParameters = function () { return [
        { type: ViewContainerRef, },
    ]; };
    return CategoryDirective;
}(ComplexBase));
export { CategoryDirective };
/**
 * Category Array Directive
 * @private
 */
var CategoriesDirective = /** @class */ (function (_super) {
    __extends(CategoriesDirective, _super);
    function CategoriesDirective() {
        return _super.call(this, 'categories') || this;
    }
    CategoriesDirective.decorators = [
        { type: Directive, args: [{
                    selector: 'e-multilevellabel>e-categories',
                    queries: {
                        children: new ContentChildren(CategoryDirective)
                    },
                },] },
    ];
    /** @nocollapse */
    CategoriesDirective.ctorParameters = function () { return []; };
    return CategoriesDirective;
}(ArrayBase));
export { CategoriesDirective };
