// @syncfusion/ej2-ng-charts/src/chart/multilevellabels.directive 2.js
var __extends = (this && this.__extends) || (function () {
    var extendStatics = Object.setPrototypeOf ||
        ({ __proto__: [] } instanceof Array && function (d, b) { d.__proto__ = b; }) ||
        function (d, b) { for (var p in b) if (b.hasOwnProperty(p)) d[p] = b[p]; };
    return function (d, b) {
        extendStatics(d, b);
        function __() { this.constructor = d; }
        d.prototype = b === null ? Object.create(b) : (__.prototype = b.prototype, new __());
    };
})();
import { Directive, ViewContainerRef, ContentChildren, ContentChild } from '@angular/core';
import { ComplexBase, ArrayBase, setValue } from '@syncfusion/ej2-ng-base';
import { CategoriesDirective } from './categories.directive';
var input = ['alignment', 'border', 'categories', 'overflow', 'textStyle'];
var outputs = ['alignment', 'border', 'categories', 'overflow', 'textStyle'];
/**
 * MultiLevelLabel Directive
 * ```html
 * <e-axis>
 * <e-multilevellabels>
 * <e-multilevellabel></e-multilevellabel>
 * </e-multilevellabels>
 * </e-axis>
 * ```
 */
var MultiLevelLabelDirective = /** @class */ (function (_super) {
    __extends(MultiLevelLabelDirective, _super);
    function MultiLevelLabelDirective(viewContainerRef) {
        var _this = _super.call(this) || this;
        _this.viewContainerRef = viewContainerRef;
        _this.tags = ['categories'];
        setValue('currentInstance', _this, _this.viewContainerRef);
        return _this;
        //this.registerEvents(outputs);
    }
    MultiLevelLabelDirective.decorators = [
        { type: Directive, args: [{
                    selector: 'e-axis>e-multilevellabels>e-multilevellabel',
                    inputs: input,
                    queries: {
                        childCategories: new ContentChild(CategoriesDirective)
                    }
                },] },
    ];
    /** @nocollapse */
    MultiLevelLabelDirective.ctorParameters = function () { return [
        { type: ViewContainerRef, },
    ]; };
    return MultiLevelLabelDirective;
}(ComplexBase));
export { MultiLevelLabelDirective };
/**
 * MultiLevelLabel Array Directive
 * @private
 */
var MultiLevelLabelsDirective = /** @class */ (function (_super) {
    __extends(MultiLevelLabelsDirective, _super);
    function MultiLevelLabelsDirective() {
        return _super.call(this, 'multilevellabels') || this;
    }
    MultiLevelLabelsDirective.decorators = [
        { type: Directive, args: [{
                    selector: 'e-axis>e-multilevellabels',
                    queries: {
                        children: new ContentChildren(MultiLevelLabelDirective)
                    },
                },] },
    ];
    /** @nocollapse */
    MultiLevelLabelsDirective.ctorParameters = function () { return []; };
    return MultiLevelLabelsDirective;
}(ArrayBase));
export { MultiLevelLabelsDirective };
