// @syncfusion/ej2-ng-charts/src/chart/selecteddataindexes.directive.js
var __extends = (this && this.__extends) || (function () {
    var extendStatics = Object.setPrototypeOf ||
        ({ __proto__: [] } instanceof Array && function (d, b) { d.__proto__ = b; }) ||
        function (d, b) { for (var p in b) if (b.hasOwnProperty(p)) d[p] = b[p]; };
    return function (d, b) {
        extendStatics(d, b);
        function __() { this.constructor = d; }
        d.prototype = b === null ? Object.create(b) : (__.prototype = b.prototype, new __());
    };
})();
import { Directive, ViewContainerRef, ContentChildren } from '@angular/core';
import { ComplexBase, ArrayBase, setValue } from '@syncfusion/ej2-ng-base';
var input = ['point', 'series'];
var outputs = ['point', 'series'];
/**
 * Selected Data Directive
 * ```html
 * <e-selecteddataindexes><e-selecteddataindex></e-selecteddataindex><e-selecteddataindexes>
 * ```
 */
var SelectedDataIndexDirective = /** @class */ (function (_super) {
    __extends(SelectedDataIndexDirective, _super);
    function SelectedDataIndexDirective(viewContainerRef) {
        var _this = _super.call(this) || this;
        _this.viewContainerRef = viewContainerRef;
        setValue('currentInstance', _this, _this.viewContainerRef);
        return _this;
        //this.registerEvents(outputs);
    }
    SelectedDataIndexDirective.decorators = [
        { type: Directive, args: [{
                    selector: 'e-selecteddataindexes>e-selecteddataindex',
                    inputs: input,
                    queries: {}
                },] },
    ];
    /** @nocollapse */
    SelectedDataIndexDirective.ctorParameters = function () { return [
        { type: ViewContainerRef, },
    ]; };
    return SelectedDataIndexDirective;
}(ComplexBase));
export { SelectedDataIndexDirective };
/**
 * SelectedDataIndex Array Directive
 * @private
 */
var SelectedDataIndexesDirective = /** @class */ (function (_super) {
    __extends(SelectedDataIndexesDirective, _super);
    function SelectedDataIndexesDirective() {
        return _super.call(this, 'selecteddataindexes') || this;
    }
    SelectedDataIndexesDirective.decorators = [
        { type: Directive, args: [{
                    selector: 'ejs-chart>e-selecteddataindexes',
                    queries: {
                        children: new ContentChildren(SelectedDataIndexDirective)
                    },
                },] },
    ];
    /** @nocollapse */
    SelectedDataIndexesDirective.ctorParameters = function () { return []; };
    return SelectedDataIndexesDirective;
}(ArrayBase));
export { SelectedDataIndexesDirective };
