// @syncfusion/ej2-ng-charts/src/chart/trendlines.directive 2.js
var __extends = (this && this.__extends) || (function () {
    var extendStatics = Object.setPrototypeOf ||
        ({ __proto__: [] } instanceof Array && function (d, b) { d.__proto__ = b; }) ||
        function (d, b) { for (var p in b) if (b.hasOwnProperty(p)) d[p] = b[p]; };
    return function (d, b) {
        extendStatics(d, b);
        function __() { this.constructor = d; }
        d.prototype = b === null ? Object.create(b) : (__.prototype = b.prototype, new __());
    };
})();
import { Directive, ViewContainerRef, ContentChildren } from '@angular/core';
import { ComplexBase, ArrayBase, setValue } from '@syncfusion/ej2-ng-base';
var input = ['animation', 'backwardForecast', 'enableTooltip', 'fill', 'forwardForecast', 'intercept', 'legendShape', 'marker', 'name', 'period', 'polynomialOrder', 'type', 'width'];
var outputs = ['animation', 'backwardForecast', 'enableTooltip', 'fill', 'forwardForecast', 'intercept', 'legendShape', 'marker', 'name', 'period', 'polynomialOrder', 'type', 'width'];
/**
 * Series Directive
 * ```html
 * <e-series-collection>
 * <e-series>
 * <e-trendlines>
 * <e-trendline>
 * </e-trendline>
 * </e-trendlines>
 * </e-series-collection>
 * ```
 */
var TrendlineDirective = /** @class */ (function (_super) {
    __extends(TrendlineDirective, _super);
    function TrendlineDirective(viewContainerRef) {
        var _this = _super.call(this) || this;
        _this.viewContainerRef = viewContainerRef;
        setValue('currentInstance', _this, _this.viewContainerRef);
        return _this;
        //this.registerEvents(outputs);
    }
    TrendlineDirective.decorators = [
        { type: Directive, args: [{
                    selector: 'e-series>e-trendlines>e-trendline',
                    inputs: input,
                    queries: {}
                },] },
    ];
    /** @nocollapse */
    TrendlineDirective.ctorParameters = function () { return [
        { type: ViewContainerRef, },
    ]; };
    return TrendlineDirective;
}(ComplexBase));
export { TrendlineDirective };
/**
 * Trendline Array Directive
 * @private
 */
var TrendlinesDirective = /** @class */ (function (_super) {
    __extends(TrendlinesDirective, _super);
    function TrendlinesDirective() {
        return _super.call(this, 'trendlines') || this;
    }
    TrendlinesDirective.decorators = [
        { type: Directive, args: [{
                    selector: 'e-series>e-trendlines',
                    queries: {
                        children: new ContentChildren(TrendlineDirective)
                    },
                },] },
    ];
    /** @nocollapse */
    TrendlinesDirective.ctorParameters = function () { return []; };
    return TrendlinesDirective;
}(ArrayBase));
export { TrendlinesDirective };
