// @syncfusion/ej2-pdf-export/src/implementation/graphics/fonts/enum 2.js
define(["require", "exports"], function (require, exports) {
    "use strict";
    Object.defineProperty(exports, "__esModule", { value: true });
    var PdfFontStyle;
    (function (PdfFontStyle) {
        PdfFontStyle[PdfFontStyle["Regular"] = 0] = "Regular";
        PdfFontStyle[PdfFontStyle["Bold"] = 1] = "Bold";
        PdfFontStyle[PdfFontStyle["Italic"] = 2] = "Italic";
        PdfFontStyle[PdfFontStyle["Underline"] = 4] = "Underline";
        PdfFontStyle[PdfFontStyle["Strikeout"] = 8] = "Strikeout";
    })(PdfFontStyle = exports.PdfFontStyle || (exports.PdfFontStyle = {}));
    var PdfFontFamily;
    (function (PdfFontFamily) {
        PdfFontFamily[PdfFontFamily["Helvetica"] = 0] = "Helvetica";
        PdfFontFamily[PdfFontFamily["Courier"] = 1] = "Courier";
        PdfFontFamily[PdfFontFamily["TimesRoman"] = 2] = "TimesRoman";
        PdfFontFamily[PdfFontFamily["Symbol"] = 3] = "Symbol";
        PdfFontFamily[PdfFontFamily["ZapfDingbats"] = 4] = "ZapfDingbats";
    })(PdfFontFamily = exports.PdfFontFamily || (exports.PdfFontFamily = {}));
    var PdfFontType;
    (function (PdfFontType) {
        PdfFontType[PdfFontType["Standard"] = 0] = "Standard";
        PdfFontType[PdfFontType["TrueType"] = 1] = "TrueType";
        PdfFontType[PdfFontType["TrueTypeEmbedded"] = 2] = "TrueTypeEmbedded";
    })(PdfFontType = exports.PdfFontType || (exports.PdfFontType = {}));
    var PdfWordWrapType;
    (function (PdfWordWrapType) {
        PdfWordWrapType[PdfWordWrapType["None"] = 0] = "None";
        PdfWordWrapType[PdfWordWrapType["Word"] = 1] = "Word";
        PdfWordWrapType[PdfWordWrapType["WordOnly"] = 2] = "WordOnly";
        PdfWordWrapType[PdfWordWrapType["Character"] = 3] = "Character";
    })(PdfWordWrapType = exports.PdfWordWrapType || (exports.PdfWordWrapType = {}));
    var PdfSubSuperScript;
    (function (PdfSubSuperScript) {
        PdfSubSuperScript[PdfSubSuperScript["None"] = 0] = "None";
        PdfSubSuperScript[PdfSubSuperScript["SuperScript"] = 1] = "SuperScript";
        PdfSubSuperScript[PdfSubSuperScript["SubScript"] = 2] = "SubScript";
    })(PdfSubSuperScript = exports.PdfSubSuperScript || (exports.PdfSubSuperScript = {}));
    var FontEncoding;
    (function (FontEncoding) {
        FontEncoding[FontEncoding["Unknown"] = 0] = "Unknown";
        FontEncoding[FontEncoding["StandardEncoding"] = 1] = "StandardEncoding";
        FontEncoding[FontEncoding["MacRomanEncoding"] = 2] = "MacRomanEncoding";
        FontEncoding[FontEncoding["MacExpertEncoding"] = 3] = "MacExpertEncoding";
        FontEncoding[FontEncoding["WinAnsiEncoding"] = 4] = "WinAnsiEncoding";
        FontEncoding[FontEncoding["PdfDocEncoding"] = 5] = "PdfDocEncoding";
        FontEncoding[FontEncoding["IdentityH"] = 6] = "IdentityH";
    })(FontEncoding = exports.FontEncoding || (exports.FontEncoding = {}));
    var TtfCmapFormat;
    (function (TtfCmapFormat) {
        TtfCmapFormat[TtfCmapFormat["Apple"] = 0] = "Apple";
        TtfCmapFormat[TtfCmapFormat["Microsoft"] = 4] = "Microsoft";
        TtfCmapFormat[TtfCmapFormat["Trimmed"] = 6] = "Trimmed";
    })(TtfCmapFormat = exports.TtfCmapFormat || (exports.TtfCmapFormat = {}));
    var TtfCmapEncoding;
    (function (TtfCmapEncoding) {
        TtfCmapEncoding[TtfCmapEncoding["Unknown"] = 0] = "Unknown";
        TtfCmapEncoding[TtfCmapEncoding["Symbol"] = 1] = "Symbol";
        TtfCmapEncoding[TtfCmapEncoding["Unicode"] = 2] = "Unicode";
        TtfCmapEncoding[TtfCmapEncoding["Macintosh"] = 3] = "Macintosh";
    })(TtfCmapEncoding = exports.TtfCmapEncoding || (exports.TtfCmapEncoding = {}));
    var TtfPlatformID;
    (function (TtfPlatformID) {
        TtfPlatformID[TtfPlatformID["AppleUnicode"] = 0] = "AppleUnicode";
        TtfPlatformID[TtfPlatformID["Macintosh"] = 1] = "Macintosh";
        TtfPlatformID[TtfPlatformID["Iso"] = 2] = "Iso";
        TtfPlatformID[TtfPlatformID["Microsoft"] = 3] = "Microsoft";
    })(TtfPlatformID = exports.TtfPlatformID || (exports.TtfPlatformID = {}));
    var TtfMicrosoftEncodingID;
    (function (TtfMicrosoftEncodingID) {
        TtfMicrosoftEncodingID[TtfMicrosoftEncodingID["Undefined"] = 0] = "Undefined";
        TtfMicrosoftEncodingID[TtfMicrosoftEncodingID["Unicode"] = 1] = "Unicode";
    })(TtfMicrosoftEncodingID = exports.TtfMicrosoftEncodingID || (exports.TtfMicrosoftEncodingID = {}));
    var TtfMacintoshEncodingID;
    (function (TtfMacintoshEncodingID) {
        TtfMacintoshEncodingID[TtfMacintoshEncodingID["Roman"] = 0] = "Roman";
        TtfMacintoshEncodingID[TtfMacintoshEncodingID["Japanese"] = 1] = "Japanese";
        TtfMacintoshEncodingID[TtfMacintoshEncodingID["Chinese"] = 2] = "Chinese";
    })(TtfMacintoshEncodingID = exports.TtfMacintoshEncodingID || (exports.TtfMacintoshEncodingID = {}));
    var FontDescriptorFlags;
    (function (FontDescriptorFlags) {
        FontDescriptorFlags[FontDescriptorFlags["FixedPitch"] = 1] = "FixedPitch";
        FontDescriptorFlags[FontDescriptorFlags["Serif"] = 2] = "Serif";
        FontDescriptorFlags[FontDescriptorFlags["Symbolic"] = 4] = "Symbolic";
        FontDescriptorFlags[FontDescriptorFlags["Script"] = 8] = "Script";
        FontDescriptorFlags[FontDescriptorFlags["Nonsymbolic"] = 32] = "Nonsymbolic";
        FontDescriptorFlags[FontDescriptorFlags["Italic"] = 64] = "Italic";
        FontDescriptorFlags[FontDescriptorFlags["ForceBold"] = 262144] = "ForceBold";
    })(FontDescriptorFlags = exports.FontDescriptorFlags || (exports.FontDescriptorFlags = {}));
    var TtfCompositeGlyphFlags;
    (function (TtfCompositeGlyphFlags) {
        TtfCompositeGlyphFlags[TtfCompositeGlyphFlags["Arg1And2AreWords"] = 1] = "Arg1And2AreWords";
        TtfCompositeGlyphFlags[TtfCompositeGlyphFlags["ArgsAreXyValues"] = 2] = "ArgsAreXyValues";
        TtfCompositeGlyphFlags[TtfCompositeGlyphFlags["RoundXyToGrid"] = 4] = "RoundXyToGrid";
        TtfCompositeGlyphFlags[TtfCompositeGlyphFlags["WeHaveScale"] = 8] = "WeHaveScale";
        TtfCompositeGlyphFlags[TtfCompositeGlyphFlags["Reserved"] = 16] = "Reserved";
        TtfCompositeGlyphFlags[TtfCompositeGlyphFlags["MoreComponents"] = 32] = "MoreComponents";
        TtfCompositeGlyphFlags[TtfCompositeGlyphFlags["WeHaveAnXyScale"] = 64] = "WeHaveAnXyScale";
        TtfCompositeGlyphFlags[TtfCompositeGlyphFlags["WeHaveTwoByTwo"] = 128] = "WeHaveTwoByTwo";
        TtfCompositeGlyphFlags[TtfCompositeGlyphFlags["WeHaveInstructions"] = 256] = "WeHaveInstructions";
        TtfCompositeGlyphFlags[TtfCompositeGlyphFlags["UseMyMetrics"] = 512] = "UseMyMetrics";
    })(TtfCompositeGlyphFlags = exports.TtfCompositeGlyphFlags || (exports.TtfCompositeGlyphFlags = {}));
});
